### PR TITLE
Fix `nan` print, simplify negative number printing.

### PR DESCRIPTION
**Pull request by drinkcat**
### test: printf: Add a test for scientific printing of negative number

This was broken before the last few commits.

### test: printf: Add nan, inf, negative zero

Add a few end-to-end tests for printf of unusual floats (nan,
infinity, negative zero).

### uucore: format: print absolute value of float, then add sign

Simplifies the code, but also fixes printing of negative and positive `NaN`:
`cargo run printf "%f %f\n" nan -nan`

Fixes part 2 of #7412.

### uucore: format: force NaN back to lowercase

Fixes formatting of `NaN` to `nan`.

Fixes part 1 of #7412.

### CODE REVIEW

File: src/uucore/src/lib/features/format/num_format.rs [hunk 1]
```diff
@@ -293,13 +290,7 @@ impl Formatter for Float {
...
         let precision = match precision {
             Some(CanAsterisk::Fixed(x)) => x,
-            None => {
-                if matches!(variant, FloatVariant::Shortest) {
-                    6
-                } else {
-                    0
-                }
-            }
+            None => 6,
```
Comment (sylvestre) [inline src/uucore/src/lib/features/format/num_format.rs:293]: could you please document what 6 means here ?

thanks

Comment (drinkcat) [reply src/uucore/src/lib/features/format/num_format.rs:293]: Sure. That's part of the commits in #7384 though (so I'll update there and rebase).

Comment (drinkcat) [reply src/uucore/src/lib/features/format/num_format.rs]: Done and rebased.

**Comment by drinkcat** [comment]
This also accidentally fixes negative number scientific printing:
```
# HEAD:
cargo run printf "%e %e\n" 1234 -1234
1.234000e+03 -1234.000000e+00
# This branch:
cargo run printf "%e %e\n" 1234 -1234
1.234000e+03 -1.234000e+03
```

I'll add a test for that, so it doesn't regress.

**Comment by drinkcat** [comment]
Rebased. Also, while doing further work I realized that I can now remove the sign handling logic from `format_float_hexadecimal` (since that only prints non-negative numbers now), so I squashed that in.
